File: 3315-construct-the-minimum-bitwise-array-ii/3315-construct-the-minimum-bitwise-array-ii.py
class Solution:
    def minBitwiseArray(self, nums: List[int]) -> List[int]:
        def solve(val: int) -> int:
            ans = float('inf')
            for i in range(32):
                num = val & ~(1 << i)
                if num < 0:
                    continue
                if (num | (num + 1) == val):
                    ans = min(ans, num)
            return -1 if ans == float('inf') else ans
        
        n = len(nums)
        ans = [-1] * n
        for i in range(n):
            ans[i] = solve(nums[i])
        return ans

File: 3315-construct-the-minimum-bitwise-array-ii/3315-construct-the-minimum-bitwise-array-ii.java
class Solution {
    public int solve(int val) {
        int ans = Integer.MAX_VALUE;
        for (int i = 0; i < 32; i++) {
            int num = val & ~(1 << i);
            if (num < 0) {
                continue;
            }
            if ((num | (num + 1)) == val) {
                ans = Math.min(ans, num);
            }
        }
        return ans == Integer.MAX_VALUE ? -1 : ans;
    }
    public int[] minBitwiseArray(List<Integer> nums) {
        int n = nums.size();
        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            ans[i] = solve(nums.get(i));
        }
        return ans;
    }
}